Clip RGB data to valid range for imshow

Zac-HD opened the pull request:
~~`Axes.imshow` can now apply the `norm` or `vmin` and `vmax` arguments to RGB images, allowing easy display of high-bit-depth data, or adjustments of brightness and contrast.~~

Out-of-range values in un-normalised RGB and RGBA images are now clipped to the nearest bound, rather than being wrapped.  This bug often hid outliers, and could make interpretation of the plotted image entirely unreliable.

Closes #9391.  Closes #5382.  Related to pydata/xarray#1796.

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature

@tacaswell, I'd appreciate any suggestions you might have about tests.

tacaswell commented:
This is definitely not going in to 2.1.2 as this is a new feature not a bug-fix.

I am not convinced that this is a good idea, if the user is passing in an RGB(A) image, then the assumption is that it is a ready-to-go image and we should do as little processing as possible on it.  Starting to do so opens up a big can of worms (for example, this uses the same norm for all of the planes, it also seems reasonable to norm each plane independently).  Once you go down this road the range of extensions seems like it will get large which is why this logic should live in the application level.

The `Normalize` methods do not return strictly [0, 1] values either, they put in sentinels for over/under/bad which needs to be accounted for.

I think the fixes should be clarifying the docstring (which seems to have already been done) and detecting when we get high-bit-depth ints and down-scaling them to 8 bit (we output Nx8 bit colors so the down sampling is going to happen eventually), and clipping the floats.

In either case, the logic needs to be `AxesImage` not in `imshow` so it applies to data later set via `im.set_data`.

jklymak commented:
Yeah, this is the same argument as in the other Issues.   RGB(A) is RGB(A).  If the upstream app doesn't know how to form an RGB array within proper limits, its hard to get excited about helping them.   I think a good argument can be made that we should clip instead of modulo values that are out of range, but I don't agree with allowing vmin/vmax.

anntzer commented:
I don't like the idea either at all, but if it does go in in some form this should probably use a machinery similar to https://github.com/matplotlib/matplotlib/pull/8738 because it'd be silly to have one mechanism for normalizing two-channel images and another for three-channel images.

dstansby commented:
I also agree on forcing RGB(A) to be between 0 and 1 (as far as I can think all it really requires is `im  = im / np.max(im)` if you're original data isn't in that range). I think we could do better on the docstring though, and erroring/warning on out of range input.

jklymak commented:
I think we can error or clip/warn/info on out-of-range for `im`.

I'm -1 on automatic normalization if values are greated than 1.  How do we know the users data doesn't really go to `>max(im)` in other images that they are trying to compare.  

I'm also not in favour of applying arbitrary user-supplied normalizations to each RGB(A) channel in `imshow`.

To me, and I assume whoever wrote `imshow`, specifying RGB(A) means the user wants that color plotted on the screen (within the limits of color gamuts etc etc).  

Its clear there is a body of users who thinks of RGB as channel1/channel2/channel3 with arbitrary data in them, and that they have developed enough intuition about what those images look like to be able to quantitatively see something in those three channels. I don't think `imshow` should change for that body of users.  On the otherhand, a new function `channelrgbimage` or whatever name makes the most sense to that community, could be easy enough to add. Then `vmin/vmax` could be specified as floats or length-3 arrays, and three normalizations (one for each channel) could be supplied to map into colorspace.  

As @anntzer points out, #8738 was designed to map 2-channel data to arbitrary colors. Thats a fair bit more general than this use case, which strictly maps to RGB.

efiring commented:
Adding to the chorus, and going a bit farther: at most, we should error out or clip with a warning.  Users who want to scale the channels are free to do so outside of mpl.  I don't see any reason why that functionality should be part of mpl.

jklymak commented:
I guess what many users want is clip, but we should `_log.warn` so unattentive users don't get a saturated image and not realize they've lost dynamic range of their image. If they don't like the warning, they can clip/normalize their data properly (or suppress the warning via `logging`.)

jklymak commented on lib/matplotlib/image.py:
```diff
@@ -13,6 +13,7 @@
 
 from math import ceil
 import os
+import warnings
```
I'd strongly suggest this is a use for `logging` instead of warnings.

```
import logging
....

_log = logging.getLogger(__name__)

....
_log.warn('Clipping...')
```

Zac-HD replied:
I disagree, but I'm happy to let @tacaswell or another maintainer settle it.

FWIW I did a search for each pattern and found 21 log calls (mostly for import errors or configuration state changes) and 66 warnings (mostly for invalid calls or data), so warnings seem to be the convention here (as in Numpy, Pandas, etc).

jklymak replied:
Logging was just added a couple of months ago.

jklymak replied:
Just to follow up, when we added logging, we decided to follow the guidelines at https://docs.python.org/3/howto/logging.html#logging-basic-tutorial for when to use which tool:

> warnings.warn() in library code if the issue is avoidable and the client application should be 
> modified to eliminate the warning
>
> logging.warning() if there is nothing the client application can do about the situation, but the 
> event should still be noted

We decided *not* go back and check that the current instances of `warnings.warn` follow that pattern, but agreed to fix them as they come along.  

I'd drop my "strongly" above, and mildly argue that this warning is the latter case, and the particular advantage of `logging` is that the warnings can be turned off if the user is OK and cognizant w/ the clipping.  I don't think `warnings.warn` can be turned off, or at least not on a  per-module basis.

Zac-HD replied:
Ah, that makes sense!  The [contributing guide is pretty good](https://matplotlib.org/devdocs/devel/contributing.html#which-logging-level-to-use), but I'll add a note with this explanation 😄

Zac-HD commented:
I'm seeing a strong consensus that matplotlib should *not* normalize or scale RGB or RGBA images, so I've dropped that part of the pull.  However this still serves to close the linked issues (as "wontfix") and allows downstream support without worrying about API compatibility.

I've moved the clipping logic into the `add_data` method where it belongs, but not added tests yet as there's ongoing design discussion.

Zac-HD commented:
Converted from `warnings.warn` to `_log.warn`, added test.

An image comparison test also seems like a good idea, but I can't get everything to build 😕.  If this is needed, would someone else be able to supply the reference images?

WeatherGod commented on doc/users/next_whats_new/2018_01_11_imshow-rgb-scaling-clipping.rst:
```diff
@@ -0,0 +1,7 @@
+When `Axes.imshow` is passed an RGB or RGBA value with out-of-range
+values, it now issues a warning and clips them to the valid range.
+The old behaviour, wrapping back in to the range, often hid outliers
+and made interpreting RGB images unreliable.
```
Should we clarify what ranges are, or perhaps link to an explanation of the ranges?

Zac-HD replied:
The ranges are explained by the docs for `Axes.imshow`, and the reference to it here is turned into a link by the backticks IIRC.

jklymak replied:
Sorry to drag this out.  I think this really needs to be in the API changes as well in case smeone was depending on the wrap behaviour:

```
Log of changes to Matplotlib that affect the outward-facing API. If updating Matplotlib breaks your scripts, this list may help you figure out what caused the breakage and how to fix it by updating your code.
```

WeatherGod commented on lib/matplotlib/image.py:
```diff
@@ -610,6 +613,23 @@ def set_data(self, A):
+            # - otherwise casting wraps extreme values, hiding outliers and
+            # making reliable interpretation impossible.
+            high = 255 if np.issubdtype(self._A.dtype, np.integer) else 1
+            if self._A.min() < 0 or high < self._A.max():
```
I wonder if we should pad this by an `eps` in case the 0-1 values came from some numerical computation, and we probably wouldn't want to spam the user with unnecessary warnings?

Zac-HD replied:
We have to clip to exactly [0 .. 1], and for consistency I would prefer to always warn if clipping.  If users feel spammed, they can either fix their data or filter the logging output.

jklymak replied:
I think the suggestion might have been to silently clip if (a-vmax)/(vmin - vmax) - 1 < 1e-10 or something that would just represent bit noise that would never affect the interpretation of the data.

Zac-HD replied:
Yes, that's the question - I just think the answer is "No, we shouldn't".

anntzer replied:
Agree that we shouldn't.

Zac-HD commented:
The remaining test errors are because the pytest `caplog` fixture is unavailable on some Travis environments, which in turn is because they are running old versions of pytest (3.1.0, caplog was added in 3.3.0, latest is 3.3.2).

I can mark this with a `skipif`, or if you don't mind requiring a newer version of pytest I'll just leave it as-is.

anntzer commented:
I would adjust .travis.yml (and whatever docs) to bump the minimal required pytest version.

jklymak commented on lib/matplotlib/image.py:
```diff
@@ -610,6 +613,23 @@ def set_data(self, A):
+            # making reliable interpretation impossible.
+            high = 255 if np.issubdtype(self._A.dtype, np.integer) else 1
+            if self._A.min() < 0 or high < self._A.max():
+                _log.warn(
```
Its `_log.warning`, isn't it?

Zac-HD replied:
Yep, turns out that `_log.warn` ["is functionally identical but deprecated"](https://docs.python.org/3/library/logging.html#logging.Logger.warning) (but not programmatically deprecated).

Zac-HD commented:
OK, I've responded to all the review comments, tests are present and passing, and Travis has been told to use a new version of pytest.

Travis *isn't* using a new version of pytest, but I think that can be fixed by blowing away some caches and re-running the job.

Zac-HD commented:
One flaky test (null byte validation), and one job that installed an old pytest again.

Zac-HD commented:
@efiring, I've done all I can to get the test passing - there are two remaining problems:

- `lib/matplotlib/tests/test_rcparams.py::test_validator_invalid` is failing with a null byte issue somewhere, which is nothing to do with this pull.  That's the Python 3.6 job on Travis.
- The Travis jobs for python 3.4 and 3.7 are still using an older (incompatible) version of pytest.  If you clear the cache are restart those jobs I would expect all the tests to pass.

efiring commented:
I confess: I know how to restart the tests, but I don't know how to clear a cache on Travis. @tacaswell knows all, and needs to approve the changes (or not) in any case, so I will leave it to him.

Zac-HD commented:
FYI it's under the "More options" button on the upper right, then "caches" and pick the relevant pull request - or just blow away everything with "delete all repository caches".

*Usually* this is all automatic, but things just haven't been expiring as they should this year :man_shrugging:

QuLogic commented:
TBH, I'm not sure about bumping the pytest requirement. I'm not seeing 3.3 packaged in a lot of distros just yet.

Zac-HD commented:
> TBH, I'm not sure about bumping the pytest requirement. I'm not seeing 3.3 packaged in a lot of distros just yet.

Given the trouble with Travis, I'm happy to ditch the check for logging - I'll just check that the array was in fact clipped to the valid range, but not that a warning was logged.

tacaswell commented:
I am also concerned about the pytest version bump.

If the cache / version issues persist try rebasing on current master.

Zac-HD commented:
@tacaswell - I'm no longer bumping the pytest version.

@efiring - tests all passing now 🎉

efiring commented:
I'm happy, but @tacaswell still has an angry red X next to his review...

Zac-HD commented:
Ping @tacaswell - if you're happy with this it can be merged; and if not I'd appreciate tips on how to improve it 😄

Zac-HD commented:
Ah, sorry - I'd missed that comment.  Now with API changes documented; and since it's been a while I also bumped the dates and rebased on master.

The [history of `credits.rst`](https://github.com/matplotlib/matplotlib/blame/master/doc/users/credits.rst) suggests that the list was generated from Git metadata (by @efiring), but that it's not automatically updated.  Unless someone knows how else it will be added, I'll therefore leave my name on the list 😄 

(and ping @tacaswell again; I think we're just waiting for your approval now)

Zac-HD commented:
Ping @jklymak and @tacaswell; I've made all the changes you wanted and would love to get this merged in time for v2.2 - the due date is tomorrow!

tacaswell commented:
Thanks!  Sorry I have been slow on review recently.
